scylla-server/src/proxy/proxy-client.ts

```diff
@@ -39,26 +61,108 @@ export default class ProxyClient {
+    try {
+      const data = JSON.parse(payload.toString()) as ServerMessage;
+      this.handleData(data);
+    } catch (error) {
+      if (error instanceof Error) {
+        console.log('Error Decoding Message: ', error.message);
+        this.handleError(error);
+      }
+    }
+  };
   /**
    * Handles receiving data from the car and:
    * 1. Logs the data
    * 2. Sends the data to the client
    * @param data The data received from Siren
    */
-  private handleData = (data: ServerMessage) => {
-    //TODO: Send data to client
-    //TODO: Log data
-    console.log('Received Data: ', data);
+  private handleData = async (data: ServerMessage) => {
+    // if first time data recieved since connecting to car
+    // then create new run
+    if (this.createNewRun) {
+      this.currentRun = await RunService.createRun(data.unix_time);
+    }
+    this.createNewRun = false;
+    // upsert the node
+    const node = await NodeService.upsertNode(data.node);
+    // start upserting data
+    if (!this.currentRun) {
+      console.log('no current run');
+      return;
+    }
+    // initializing params
+    let latitude = undefined;
+    let longitude = undefined;
+    let radius = undefined;
+    let locationName = undefined;
+    let driverName = undefined;
+    let systemName = undefined;
+
+    // iterating and upserting
+    const clientData: ClientData[] = [];
+    for (const serverdata of data.data) {
+      // transform serverdata into client data to send to ProxyServer
+      const clientdata: ClientData = {
+        name: serverdata.name,
+        value: serverdata.value,
+        units: serverdata.units,
+        timestamp: data.unix_time
+      };
+      clientData.push(clientdata);
+
+      switch (serverdata.name) {
+        case 'driverUser':
+          driverName = serverdata.value as unknown as string;
+          break;
+        case 'systemName':
+          systemName = serverdata.value as unknown as string;
+          break;
+        case 'locationName':
+          locationName = serverdata.value as unknown as string;
+          break;
+        case 'latitude':
+          latitude = serverdata.value as number;
+          break;
+        case 'longitude':
+          longitude = serverdata.value as number;
+          break;
+        case 'radius':
+          radius = serverdata.value as number;
+          break;
+        default:
+          await DataTypeService.upsertDataType(serverdata.name, serverdata.units, node.name);
+          await DataService.addData(serverdata, data.unix_time, serverdata.value as number, this.currentRun.id);
+      }
+    }
+
+    if (systemName) {
+      SystemService.upsertSystem(systemName, this.currentRun.id);
+    }
+    if (driverName) {
+      DriverService.upsertDriver(driverName, this.currentRun.id);
+    }
+    if (latitude && longitude && radius && locationName) {
+      LocationService.upsertLocation(locationName, latitude, longitude, radius, this.currentRun.id);
```

Comment:
should also be continuously tracking latitude/ longitude as a dataType - Peyton

Reply:
can u elaborate on this a lil, i made it so that latitude and longitude are upserted as dataTypes